### ISSUE TITLE
`associated_type_bounds`: errors with ambiguous associated type

### ISSUE THREAD

**Issue by KrishnaSannasi** [rust-lang/rust#61752]
I think I found a bug with associated type bounds! #52662

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=28d566e7244a0b3ad21b73e31418894d
```rust
#![feature(associated_type_bounds)]

trait Foo {
    type Bar;
}

fn works<F: Foo>() where F::Bar: Copy {}

// This should be fine, but I get `error[E0221]` instead
fn breaks<F: Foo<Bar: Foo>>() where F::Bar: Copy {}
```
This is the exact same error you get when you do this
```rust
trait Foo<T> {
    type Bar;
}

fn breaks<F: Foo<String> + Foo<u32>>() where F::Bar: Copy {}
```
Because Rust can't tell which `Foo::<_>::Bar` you mean, but in the case of `associated_type_bounds`, there is no ambiguity, which makes this error confusing.

**Comment by Centril** [comment]
Seems like a bug indeed!

cc @alexreg @nikomatsakis

**Comment by Centril** [comment]
Same error with:

```rust
fn breaks2<F>() where F: Foo<Bar: Foo>, F::Bar: Copy {}
```

Although the following results in the same problem:

```rust
fn breaks3_no_atb<F>() where F: Foo, Foo::Bar: Foo, F::Bar: Copy {}
```

so it may not be a bug after all?

**Comment by KrishnaSannasi** [comment]
Ok, maybe just the wrong error in the desugaring?

**Comment by Centril** [comment]
@KrishnaSannasi Oops; I wrote `Foo::Bar: Foo`; should be `F::Bar: Foo`. Sorry!

**Comment by alexreg** [comment]
It's a very weird bug yes, confirmed. For now, just avoid splitting up your associated type bounds like this, but I'll try to get it fixed soon.

**Comment by alexreg** [comment]
@KrishnaSannasi Hopefully you can see above a fix PR has been submitted now. It seems to do the job, but will need review then merge...
